src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.MoveFileCodeAction.cs

```diff
@@ -0,0 +1,187 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings.SyncNamespace
+{
+    internal abstract partial class AbstractSyncNamespaceService<TNamespaceDeclarationSyntax, TCompilationUnitSyntax>
+        where TNamespaceDeclarationSyntax : SyntaxNode
+        where TCompilationUnitSyntax : SyntaxNode 
+    {
+        internal sealed class MoveFileCodeAction : CodeAction
+        {
+            private readonly State _state;
+            private readonly ImmutableArray<string> _newfolders;
+
+            public override string Title
+                => _newfolders.Length > 0 
+                ? string.Format(FeaturesResources.Move_file_to_0_folder, string.Join(PathUtilities.DirectorySeparatorStr, _newfolders)) 
```

Comment:
i'm curious what this looks like in practice.  can you show with some Roslyn deeply nested folder case?

Reply:
I would remove the 'folder' part.  It feels pretty redundant. :)

Reply:
Done in this rebased PR
https://github.com/dotnet/roslyn/pull/30920